### PR TITLE
thread: add CPython 3.14 ExceptHookArgs

### PR DESCRIPTION
**Pull request by youknowone**
`_thread._excepthook` existed as a no-op stub while `_thread._ExceptHookArgs` did not exist
at all. `threading.py` imports the two together:

```python
from _thread import _excepthook as excepthook, _ExceptHookArgs as ExceptHookArgs
```

so the missing name raised `ImportError` and `threading` fell back to its pure-Python
implementation for both: `threading.ExceptHookArgs` was a function wrapping a namedtuple
rather than a type, and calling `_thread._excepthook` directly returned `None` and wrote
nothing. Measured against both oracles — CPython 3.14 has both names natively, PyPy has
neither and uses the fallback throughout. pyre was the only one carrying a stub that claims
the entry point and then does nothing with it.

- `_thread._ExceptHookArgs`: the immutable, non-baseable struct sequence with fields
  `exc_type` / `exc_value` / `exc_traceback` / `thread`, built on the existing `_structseq`
  port with `acceptable_as_base_class` cleared.
- `_thread._excepthook`: ports `thread_excepthook` — rejects a non-`ExceptHookArgs`
  argument, returns silently for `SystemExit`, resolves the stream from `sys.stderr` and
  falls back to the thread's saved `_stderr`, writes the `Exception in thread <name>:`
  banner (the native ident when the args carry no thread), and flushes.
- `error::write_exception_from_parts`: the `_PyErr_Display(file, exc_type, exc_value,
  exc_tb)` shape. The hook receives the traceback as a separate structseq field, so it has
  to render that value rather than reading `exc_value.__traceback__`; the existing walk is
  split into `write_traceback_chain_from_tb`, which takes a traceback directly.

With both names present `threading.py` takes the native path, so
`threading.ExceptHookArgs is _thread._ExceptHookArgs` and
`threading.excepthook is _thread._excepthook`.

## Validation

- `python3 scripts/extract-llbc.py` + `cargo build --release --features dynasm`
- `python3 pyre/extra_tests/run.py --dynasm-only --filter builtin_thread -v` — 1/1
- the new `builtin_thread.py` snippet also passes unmodified under CPython 3.14
- `python3 pyre/check.py --synthetic-only --backend dynasm target/release/pyre-dynasm` —
  311 passed, `synth/inline_chain_depth_typeflip` reported a >20s timeout. That timeout is
  machine contention, not this change: the bench passes on its own at 5.60s against a 20s
  budget, and a binary built from `34596313f5` (the commit that added the bench) spends the
  same 5.7s of user CPU on it under the same load.

## Pre-existing on this base, not from this branch

`test.test_threading` segfaults part-way through (rc=139) on base `0f2775fd94` with freshly
extracted LLBC. Reverting this branch's two source files and rebuilding on the same base
reproduces the crash at the same point, so it is not caused by this change. A binary built
from `34596313f5` completes the suite (`Ran 228 tests`, one failure,
`test_stop_the_world_during_finalization`), which puts the change between `34596313f5` and
`0f2775fd94` — either `9c808f9e2d` / `0f2775fd94` themselves or the re-extracted LLBC that
those commits' prepass rework produces.

## Summary by CodeRabbit

* **New Features**
  * Added CPython 3.14-compatible uncaught thread exception handling via `threading.excepthook`/`_thread._excepthook`, including strict `_ExceptHookArgs` validation and automatic flushing to the chosen stderr.
  * Thread excepthook now emits CPython-style “Exception in thread …” headers with full exception text.
* **Bug Fixes**
  * Improved structured exception rendering: better cause/context handling, `ExceptionGroup` truncation, `SyntaxError` formatting, and more robust `__notes__`/traceback output.
* **Tests**
  * Expanded runtime coverage for `_thread` argument behavior, formatting edge cases, and stderr/sys fallback behavior.

### CODE REVIEW

File: pyre/pyre-interpreter/src/error.rs [hunk 1]
```diff
@@ -1468,6 +1468,46 @@ pub fn write_exception<W: Write>(
...
+    // `PyErr_DisplayException`: older explicit cause/context entries precede
+    // the current value, while the current traceback comes from the argument
+    // supplied by ExceptHookArgs.
+    write_chained_context(writer, exc_value)?;
+    if !exc_tb.is_null()
+        && !unsafe { pyre_object::is_none(exc_tb) }
+        && unsafe { crate::pytraceback::is_pytraceback(exc_tb) }
+    {
+        writeln!(writer, "Traceback (most recent call last):")?;
+        write_traceback_chain_from_tb(writer, exc_tb)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in 0eb06ab050: the supplied traceback is installed only when the exception has no resident traceback, and display then reads the exception-owned traceback. Covered by the existing-vs-supplied traceback assertions.

File: pyre/pyre-interpreter/src/module/thread/mod.rs [hunk 2]
```diff
@@ -1670,6 +1670,201 @@ fn interrupt_main(args: &[PyObjectRef]) -> Result<PyObjectRef, crate::PyError> {
...
+    let thread_slot = pin_root_slot(thread);
+
+    // `PyFile_WriteString("Exception in thread ", file)`.
+    thread_excepthook_write(file_slot, w_str_new("Exception in thread "))?;
+
+    // `PyObject_GetOptionalAttr(thread, "name")`; a missing name falls back
+    // to the native thread identifier, while a raising descriptor propagates.
+    let thread = pyre_object::gc_roots::shadow_stack_get(thread_slot);
+    let name = if !unsafe { is_none(thread) } {
+        crate::baseobjspace::findattr_result(thread, "name")?
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/module/thread/mod.rs]: Fixed in 0eb06ab050: optional attribute lookup now suppresses only AttributeError, so a NameError raised by the name descriptor propagates. Regression coverage is included.

File: pyre/pyre-interpreter/src/module/thread/mod.rs [hunk 3]
```diff
@@ -1670,6 +1670,201 @@ fn interrupt_main(args: &[PyObjectRef]) -> Result<PyObjectRef, crate::PyError> {
...
+    };
+    let exc_value_slot = pin_root_slot(exc_value);
+    let exc_traceback_slot = pin_root_slot(exc_traceback);
+    let thread_slot = pin_root_slot(thread);
+
+    // `_PySys_GetOptionalAttr("stderr")`; when it is absent/None, use the
+    // Thread object's saved `_stderr`, unless both the stream and thread are
+    // None.
+    let sys = crate::importing::get_sys_module("sys")
+        .ok_or_else(|| crate::PyError::runtime_error("sys module is unavailable"))?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/module/thread/mod.rs]: Fixed in 0eb06ab050: stderr is read from the interpreter-owned sys registry rather than the replaceable sys.modules entry. The test replaces sys.modules["sys"] and verifies the real stream wins.

File: pyre/pyre-interpreter/src/error.rs [hunk 4]
```diff
@@ -1468,6 +1468,46 @@ pub fn write_exception<W: Write>(
...
+    }
+    writeln!(writer, "{}", render_exc_object(exc_value))?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in 0eb06ab050: rendering calls the actual exception __str__, preserving KeyError, OSError, custom conversion, and the conversion-failure placeholder.

File: pyre/pyre-interpreter/src/error.rs [hunk 5]
```diff
@@ -1468,6 +1468,46 @@ pub fn write_exception<W: Write>(
...
+            return writeln!(writer, "NoneType: None");
+        }
+        let value_type = crate::typedef::r#type(exc_value)
+            .map(|tp| unsafe { pyre_object::w_type_get_name(tp.as_ptr()).to_string() })
+            .unwrap_or_else(|| "NULL".to_string());
+        return writeln!(
+            writer,
+            "TypeError: print_exception(): Exception expected for value, {value_type} found"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:1516]: Kept intentionally after verifying CPython 3.14.2 at runtime: _thread._excepthook with exc_value=None prints exactly "NoneType: None"; an int produces the generic TypeError. Both cases now have explicit regression assertions.

File: pyre/pyre-interpreter/src/error.rs [hunk 6]
```diff
@@ -1468,6 +1468,46 @@ pub fn write_exception<W: Write>(
...
+    if !exc_tb.is_null()
+        && !unsafe { pyre_object::is_none(exc_tb) }
+        && unsafe { crate::pytraceback::is_pytraceback(exc_tb) }
+    {
+        writeln!(writer, "Traceback (most recent call last):")?;
+        write_traceback_chain_from_tb(writer, exc_tb)?;
+    }
+    writeln!(writer, "{}", render_exc_object(exc_value))?;
+    write_exception_notes(writer, exc_value)
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in f90164bff7: object-backed SyntaxError values use the structured source/caret renderer and raw msg while retaining traceback, chain, and notes ordering.

File: pyre/pyre-interpreter/src/module/thread/mod.rs [hunk 7]
```diff
@@ -1670,6 +1670,201 @@ fn interrupt_main(args: &[PyObjectRef]) -> Result<PyObjectRef, crate::PyError> {
...
+            let ns = pyre_object::w_type_get_dict_ptr(ty) as PyObjectRef;
+            pyre_object::w_dict_setitem_str_no_proxy(
+                ns,
+                "__doc__",
+                pyre_object::gc_roots::shadow_stack_get(doc_slot),
+            );
+        }
+        pyre_object::gc_roots::shadow_stack_get(ty_slot) as usize
+    }) as PyObjectRef
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/module/thread/mod.rs:1702]: No additional root is required here: w_type_new_builtin allocates builtin types through the documented Box-immortal malloc_typed path and registers builtin type roots; OnceLock stores that stable immortal address. A temporary root only protects construction-time allocations.

File: pyre/pyre-interpreter/src/error.rs [hunk 8]
```diff
@@ -1468,6 +1468,68 @@ pub fn write_exception<W: Write>(
...
+        unsafe {
+            pyre_object::interp_exceptions::w_exception_set_traceback(exc_value, exc_tb);
+        }
+    }
+
+    // Older explicit cause/context entries precede the current value.  The
+    // current traceback is the exception-resident value after the conditional
+    // installation above, exactly as `print_exception_recursive` observes it.
+    let exc_value = pyre_object::gc_roots::shadow_stack_get(exc_value_slot);
+    write_chained_context(writer, exc_value)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in 33aa9ca446. The renderer now shares the same visited identity set across cause/context and ExceptionGroup traversal as lib-python/3/traceback.py. Self-cause and two-node context-cycle regressions verify each exception is emitted once.

File: pyre/pyre-interpreter/src/error.rs [hunk 9]
```diff
@@ -1468,6 +1468,68 @@ pub fn write_exception<W: Write>(
...
+    let exc_value = pyre_object::gc_roots::shadow_stack_get(exc_value_slot);
+    let current_tb =
+        unsafe { pyre_object::interp_exceptions::w_exception_get_traceback(exc_value) };
+    if !current_tb.is_null() {
+        writeln!(writer, "Traceback (most recent call last):")?;
+        write_traceback_chain_from_tb(writer, current_tb)?;
+    }
+    let exc_value = pyre_object::gc_roots::shadow_stack_get(exc_value_slot);
+    writeln!(writer, "{}", render_exc_object(exc_value))?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed across f90164bff7 and 33aa9ca446. SyntaxError uses structured source/caret output, and ExceptionGroup now follows the Python 3.14 traceback formatter group-depth, width, indentation, nested-leaf, and closing-frame structure. The common two-leaf output matches CPython byte-for-byte.

File: pyre/pyre-interpreter/src/error.rs [hunk 10]
```diff
@@ -1529,11 +1585,234 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+    }
+    if unsafe { pyre_object::interp_exceptions::w_exception_get_kind(exc) } == ExcKind::SyntaxError
+    {
+        write_syntax_error_object(&mut rendered, exc)?;
+    } else {
+        let header = render_exc_object_wtf8(exc);
+        rendered.write_all(header.as_bytes())?;
+        rendered.write_all(b"\n")?;
+    }
+    write_exception_notes(&mut rendered, exc)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in f37c33971a. Every chain node, group tuple, child, note container, and note item is held in a shadow-stack slot and reloaded after Python calls. The regression forces gc.collect() from __str__ and then reads notes and later group children.

File: pyre/extra_tests/snippets/builtin_thread.py [hunk 11]
```diff
@@ -1,3 +1,293 @@
...
+
+group_output = hook_output(
+    ExceptionGroup(
+        "outer",
+        [
+            ValueError("first leaf"),
+            ExceptionGroup("inner", [TypeError("nested leaf")]),
+        ],
+    )
+)
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/extra_tests/snippets/builtin_thread.py:305]: No project-wide Ruff target change is needed for this PR. The repository pre-commit check passes this file, and the snippet is deliberately executed under the project target, CPython 3.14. Changing global Ruff semantics to address a third-party advisory false positive would broaden this focused runtime fix.

File: pyre/pyre-interpreter/src/error.rs [hunk 12]
```diff
@@ -1529,11 +1585,234 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+        write_exception_object_recursive(writer, older, context)?;
+        emit_exception_text(writer, banner.as_bytes(), context, b'|')?;
+    }
+
+    if exception_group_items(exc).is_some() {
+        write_exception_group(writer, exc, context)
+    } else {
+        write_plain_exception_object(writer, exc, context)
+    }
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:1707]: Fixed in f37c33971a by matching TracebackException capture with an explicit chain work vector rather than native recursion. A 1,200-node context chain now renders completely under both CPython 3.14 and pyre.

File: pyre/pyre-interpreter/src/error.rs [hunk 13]
```diff
@@ -1529,11 +1585,234 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+    if let Some((older, banner)) = chained {
+        write_exception_object_recursive(writer, older, context)?;
+        emit_exception_text(writer, banner.as_bytes(), context, b'|')?;
+    }
+
+    if exception_group_items(exc).is_some() {
+        write_exception_group(writer, exc, context)
+    } else {
+        write_plain_exception_object(writer, exc, context)
+    }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in f37c33971a. Group detection and exceptions lookup now run once; the resulting rooted child-slot vector is passed directly into write_exception_group.

File: pyre/pyre-interpreter/src/error.rs [hunk 14]
```diff
@@ -1529,11 +1585,234 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+                writer,
+                format!("and {remaining} more exception{suffix}\n").as_bytes(),
+                context,
+                b'|',
+            )?;
+        } else {
+            let child = exceptions[index];
+            context.seen.insert(child as usize);
+            write_exception_object_recursive(writer, child, context)?;
+        }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:1843]: Fixed in f37c33971a. The group tuple is rooted first, every child is pinned into its own shadow-stack slot, and each child is reloaded immediately before recursive formatting. The test forces collection while formatting the first child and verifies later nested children remain intact.

File: pyre/pyre-interpreter/src/importing.rs [hunk 15]
```diff
@@ -1444,6 +1444,23 @@ pub fn get_sys_module(name: &str) -> Option<PyObjectRef> {
...
+/// module registry and is independently walked as a GC root, so its original
+/// `sys` entry is the corresponding owner.
+pub fn get_interpreter_sys_module() -> Option<PyObjectRef> {
+    SYS_MODULES
+        .lock()
+        .unwrap()
+        .get("sys")
+        .copied()
+        .map(|module| module as PyObjectRef)
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/importing.rs:1462]: Fixed in f37c33971a by recovering the poisoned SYS_MODULES guard with into_inner(), consistent with the exception-class registry.

File: pyre/pyre-interpreter/src/error.rs [hunk 16]
```diff
@@ -1608,49 +1966,63 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+    let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
     if exc.is_null() || !unsafe { pyre_object::is_exception(exc) } {
-        return String::from("<no exception>");
+        return Wtf8Buf::from_string("<no exception>".to_string());
     }
     let name = exc_object_class_name(exc).unwrap_or_else(|| {
+        let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
         let kind = unsafe { pyre_object::interp_exceptions::w_exception_get_kind(exc) };
         exc_kind_name(kind).to_string()
     });
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:2083]: Fixed in 87f39d417c. Heap exception types now use __qualname__ and prefix __module__ except for builtins/__main__, with nested and module-qualified regression coverage against CPython 3.14.

File: pyre/pyre-interpreter/src/error.rs [hunk 17]
```diff
@@ -1529,11 +1592,288 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+    if unsafe { pyre_object::interp_exceptions::w_exception_get_kind(exc) } == ExcKind::SyntaxError
+    {
+        write_syntax_error_object(&mut rendered, exc)?;
+    } else {
+        let header = render_rooted_exc_object_wtf8(exc_slot);
+        rendered.write_all(header.as_bytes())?;
+        rendered.write_all(b"\n")?;
+    }
+    let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
+    write_exception_notes(&mut rendered, exc)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:1731]: Fixed in 87f39d417c. List/tuple note containers render each item, while other non-None values use repr, with tuple and scalar regression coverage against CPython 3.14.

File: pyre/pyre-interpreter/src/error.rs [hunk 18]
```diff
@@ -1529,11 +1606,288 @@ pub fn eprint_syntax_error(err: &PyError) {
...
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:1731]: Fixed in 985530578f. Missing __notes__ still behaves as absent, while a non-AttributeError raised by a custom descriptor is rooted, repr-rendered, and emitted with CPython 3.14's synthetic-note diagnostic. Regression coverage uses a raising descriptor.

File: pyre/pyre-interpreter/src/error.rs [hunk 19]
```diff
@@ -1529,11 +1606,288 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+    if !tb.is_null() {
+        writeln!(rendered, "Traceback (most recent call last):")?;
+        write_traceback_chain_from_tb(&mut rendered, tb)?;
+    }
+    let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
+    if unsafe { pyre_object::interp_exceptions::w_exception_get_kind(exc) } == ExcKind::SyntaxError
+    {
+        write_syntax_error_object(&mut rendered, exc)?;
+    } else {
+        let header = render_rooted_exc_object_wtf8(exc_slot);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:1726]: Fixed in 985530578f. The native formatter now ports the Python 3.14 candidate limits and weighted edit distance, gathers NameError frame namespaces, AttributeError object names, and ImportError module names, and adds the stdlib import hint. NameError, AttributeError, and ImportError regressions match CPython 3.14.

File: pyre/pyre-interpreter/src/error.rs [hunk 20]
```diff
@@ -1529,11 +1606,288 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+    exc_slot: usize,
+    context: &ExceptionPrintContext,
+) -> std::io::Result<()> {
+    let mut rendered = Vec::new();
+    let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
+    let tb = unsafe { pyre_object::interp_exceptions::w_exception_get_traceback(exc) };
+    if !tb.is_null() {
+        writeln!(rendered, "Traceback (most recent call last):")?;
+        write_traceback_chain_from_tb(&mut rendered, tb)?;
+    }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:1720]: Fixed in 985530578f. write_traceback_chain_from_tb pins every current traceback before formatting, reloads it from the shadow-stack slot after formatting, and only then reads w_next, so moving collections cannot leave a stale chain pointer.

File: pyre/pyre-interpreter/src/error.rs [hunk 21]
```diff
@@ -1583,74 +1944,368 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+            continue;
+        }
+        let candidate_len = candidate.chars().count();
+        let mut max_distance = (candidate_len + wrong_len + 3) * SUGGESTION_MOVE_COST / 6;
+        max_distance = max_distance.min(best_distance.saturating_sub(1));
+        let distance = suggestion_distance(wrong_name, candidate, max_distance);
+        if distance <= max_distance && (suggestion.is_none() || distance < best_distance) {
+            suggestion = Some(candidate.clone());
+            best_distance = distance;
+        }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:2201]: Fixed in b218ac430c. The candidate threshold now clamps to the current best distance exactly, preserving CPython's acceptance of the first equal-distance candidate.

File: pyre/pyre-interpreter/src/error.rs [hunk 22]
```diff
@@ -1583,74 +1944,368 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+    while !current.is_null() && unsafe { crate::pytraceback::is_pytraceback(current) } {
+        let slot = pyre_object::gc_roots::shadow_stack_len();
+        pyre_object::gc_roots::pin_root(current);
+        last_slot = Some(slot);
+        let rooted_current = pyre_object::gc_roots::shadow_stack_get(slot);
+        current = unsafe { crate::pytraceback::w_pytraceback_get_w_next(rooted_current) };
+    }
+    let tb = pyre_object::gc_roots::shadow_stack_get(last_slot?);
+    let frame = unsafe { crate::pytraceback::w_pytraceback_get_frame(tb) };
+    (!frame.is_null()).then_some(frame)
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in b218ac430c. The traceback head is pinned once and the chain is traversed without allocating, so deep tracebacks consume one shadow-stack slot while the rooted head keeps the complete chain live.

File: pyre/pyre-interpreter/src/error.rs [hunk 23]
```diff
@@ -1583,74 +1944,368 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+                dict_string_keys(locals, &mut names);
+                if let Some(self_obj) = unsafe { pyre_object::w_dict_getitem_str(locals, "self") } {
+                    if crate::baseobjspace::getattr_str(self_obj, &wrong_name).is_ok() {
+                        return Some(format!(". Did you mean: 'self.{wrong_name}'?"));
+                    }
+                }
+            }
+            dict_string_keys(frame.get_w_globals(), &mut names);
+            dict_string_keys(frame.fget_f_builtins(), &mut names);
+            best_suggestion(&names, &wrong_name)
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:2389]: Fixed in b218ac430c. NameError candidate collection now anchors the traceback frame with FrameAnchor, reloads it after allocating calls, pins locals and self across user-code lookup, and re-derives the live frame before globals and builtins reads.

File: pyre/pyre-interpreter/src/error.rs [hunk 24]
```diff
@@ -1583,74 +1944,368 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+    pyre_object::gc_roots::pin_root(notes);
+    let notes = pyre_object::gc_roots::shadow_stack_get(notes_slot);
+    let is_list = unsafe { crate::baseobjspace::isinstance_list_w(notes) };
+    let is_tuple = unsafe {
+        pyre_object::is_tuple(notes)
+            || crate::typedef::gettypefor(&pyre_object::TUPLE_TYPE).is_some_and(|tuple_type| {
+                crate::baseobjspace::isinstance_w(notes, tuple_type.as_ptr())
+            })
+    };
+    if is_list || is_tuple {
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in 91cf6f9e77. Exception notes now recognize non-string collections.abc.Sequence values and render each item, with a range regression test.

File: pyre/pyre-interpreter/src/error.rs [hunk 25]
```diff
@@ -1583,74 +1944,368 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+    let wrong_name = unsafe { pyre_object::w_str_get_value(wrong) }.to_string();
+    let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
+    let tb = unsafe { pyre_object::interp_exceptions::w_exception_get_traceback(exc) };
+    let suggestion = match kind {
+        ExcKind::AttributeError => {
+            let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
+            let obj = unsafe { pyre_object::interp_exceptions::w_exception_get_attr_obj(exc) };
+            object_dir_strings(obj).and_then(|mut names| {
+                if !wrong_name.starts_with('_') {
+                    names.retain(|name| !name.starts_with('_'));
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:2335]: Fixed in 91cf6f9e77. AttributeError suggestions now preserve private candidates when the failing object is self in the innermost traceback frame, with a regression test.

File: pyre/pyre-interpreter/src/error.rs [hunk 26]
```diff
@@ -1583,74 +1944,447 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+                }
+            }
         }
-        let s = unsafe { crate::display::py_str(item) }
-            .unwrap_or_else(|_| "<unprintable note>".to_string());
-        for line in s.lines() {
-            writeln!(writer, "{}", line)?;
+        for item_slot in item_slots {
+            let item = pyre_object::gc_roots::shadow_stack_get(item_slot);
+            let s = unsafe { crate::display::py_str(item) }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in eb5f3f0ac5. Exception notes now render through py_str_wtf8 and write WTF-8 bytes directly, with a lone-surrogate regression test.

File: pyre/pyre-interpreter/src/error.rs [hunk 27]
```diff
@@ -1583,74 +1944,447 @@ fn write_exception_notes<W: Write>(writer: &mut W, exc: PyObjectRef) -> std::io:
...
+    let suggestion = match kind {
+        ExcKind::AttributeError => {
+            let exc = pyre_object::gc_roots::shadow_stack_get(exc_slot);
+            let obj = unsafe { pyre_object::interp_exceptions::w_exception_get_attr_obj(exc) };
+            let obj_slot = pyre_object::gc_roots::shadow_stack_len();
+            pyre_object::gc_roots::pin_root(obj);
+            let obj = pyre_object::gc_roots::shadow_stack_get(obj_slot);
+            object_dir_strings(obj).and_then(|mut names| {
+                let obj = pyre_object::gc_roots::shadow_stack_get(obj_slot);
+                if !wrong_name.starts_with('_') && !traceback_self_is(tb, obj) {
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs:2334]: Fixed in eb5f3f0ac5. The traceback is pinned before invoking object.__dir__ and reloaded afterward, with a moving-GC __dir__ regression test.

File: pyre/pyre-interpreter/src/error.rs [hunk 28]
```diff
@@ -1529,11 +1606,288 @@ pub fn eprint_syntax_error(err: &PyError) {
...
+                writer,
+                format!("and {remaining} more exception{suffix}\n").as_bytes(),
+                context,
+                b'|',
+            )?;
+        } else {
+            let child = pyre_object::gc_roots::shadow_stack_get(exceptions[index]);
+            context
+                .seen
+                .insert(pyre_object::gc_hook::gc_identity_hash(child as usize));
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/error.rs]: Fixed in eb5f3f0ac5. Displayed ExceptionGroup children are registered in seen before any child chain is traversed, with a sibling-cause regression test.
